// @kosko/require/dist/index.node.mjs
import { createRequire } from 'node:module';
import { stat } from 'node:fs/promises';
import { resolve as resolve$1, join } from 'node:path';
import { cwd, env, version } from 'node:process';
import { getErrorCode } from '@kosko/common-utils';
import { pathToFileURL } from 'node:url';

let BASE_EXTENSIONS = [
    ".js",
    ".json"
];
function getRequireExtensions() {
    {
        let req = createRequire(import.meta.url), extensions = new Set([
            ".cjs",
            ".mjs",
            ...BASE_EXTENSIONS,
            ...req("../lib/node-extensions.cjs")()
        ]);
        return [
            ...extensions
        ];
    }
}

async function fileExists(path) {
    try {
        let stats = await stat(path);
        return stats.isFile();
    } catch (err) {
        if ("ENOENT" === getErrorCode(err)) return !1;
        throw err;
    }
}
async function resolve(id, options = {}) {
    let { baseDir =cwd() , extensions =getRequireExtensions()  } = options, resolved = resolve$1(baseDir, id), index = join(resolved, "index"), possiblePaths = [
        resolved,
        ...extensions.map((ext)=>resolved + ext),
        ...extensions.map((ext)=>index + ext)
    ];
    for (let path of possiblePaths)if (await fileExists(path)) return path;
}

let ESM_IMPORT_DISABLED = "1" === env.ESM_IMPORT_DISABLED;
async function importDefault(id) {
    let mod = await import(id);
    return mod.default;
}
let importWithOpts = (async ()=>{
    if (!ESM_IMPORT_DISABLED && function() {
        let [major, minor] = version.substring(1).split(".").map((x)=>parseInt(x, 10));
        switch(major){
            case 16:
                return minor >= 14;
            case 17:
                return minor >= 1;
            default:
                return major >= 18;
        }
    }()) return importDefault("../lib/import-with-opts.mjs");
})(), req = createRequire(import.meta.url);
function requireModule(path) {
    let mod = req(path);
    return mod && mod.__esModule ? mod : {
        default: mod
    };
}
async function tryImport(path, fn) {
    let url = pathToFileURL(path).toString();
    try {
        return await fn(url);
    } catch (err) {
        if ("ERR_UNKNOWN_FILE_EXTENSION" !== getErrorCode(err)) throw err;
    }
    return requireModule(path);
}
async function importJson(path) {
    let importFn = await importWithOpts;
    return importFn ? tryImport(path, (url)=>importFn(url, {
            assert: {
                type: "json"
            }
        })) : requireModule(path);
}
async function importPath(path) {
    return ESM_IMPORT_DISABLED ? requireModule(path) : path.endsWith(".json") ? importJson(path) : tryImport(path, (url)=>import(url));
}

export { getRequireExtensions, importPath, resolve };
