// @kosko/log/dist/index.node.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var pc = require('picocolors');
var cleanStack = require('clean-stack');
var stringify = require('fast-safe-stringify');
var commonUtils = require('@kosko/common-utils');
var node_process = require('node:process');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var pc__default = /*#__PURE__*/_interopDefault(pc);
var cleanStack__default = /*#__PURE__*/_interopDefault(cleanStack);
var stringify__default = /*#__PURE__*/_interopDefault(stringify);

exports.LogLevel = void 0;
!function(LogLevel) {
    LogLevel[LogLevel.Trace = 1] = "Trace", LogLevel[LogLevel.Debug = 2] = "Debug", LogLevel[LogLevel.Info = 3] = "Info", LogLevel[LogLevel.Warn = 4] = "Warn", LogLevel[LogLevel.Error = 5] = "Error", LogLevel[LogLevel.Fatal = 6] = "Fatal";
}(exports.LogLevel || (exports.LogLevel = {}));
let LOG_LEVEL_MAP = {
    trace: exports.LogLevel.Trace,
    debug: exports.LogLevel.Debug,
    info: exports.LogLevel.Info,
    warn: exports.LogLevel.Warn,
    error: exports.LogLevel.Error,
    fatal: exports.LogLevel.Fatal
};
function logLevelFromString(level) {
    return LOG_LEVEL_MAP[level];
}

class Logger {
    log(level, message, options = {}) {
        if (this.enabled(level)) {
            let { time =new Date() , ...rest } = options;
            this.writer.write({
                loggerLevel: this.level,
                level,
                message,
                time,
                ...rest
            });
        }
    }
    getLevel() {
        return this.level;
    }
    setLevel(level) {
        this.level = level;
    }
    enabled(level) {
        return this.level <= level;
    }
    getWriter() {
        return this.writer;
    }
    setWriter(writer) {
        this.writer = writer;
    }
    constructor(options){
        this.level = options.level, this.writer = options.writer;
    }
}
function createLoggerFactory(createWriter) {
    return (options = {})=>{
        let { level =exports.LogLevel.Info , writer =createWriter()  } = options;
        return new Logger({
            level,
            writer
        });
    };
}

let COLOR_MAP = {
    [exports.LogLevel.Trace]: pc__default.default.gray,
    [exports.LogLevel.Debug]: pc__default.default.cyan,
    [exports.LogLevel.Info]: pc__default.default.green,
    [exports.LogLevel.Warn]: pc__default.default.yellow,
    [exports.LogLevel.Error]: pc__default.default.red,
    [exports.LogLevel.Fatal]: pc__default.default.bgRed
}, LEVEL_MAX_LENGTH = Math.max(...Object.keys(exports.LogLevel).map((x)=>x.length));
function formatData(data) {
    return stringify__default.default(data, void 0, "  ");
}
class NodeLogWriter {
    write(log) {
        let { level , time , message , error , data  } = log, content = `${function(level) {
            let label = exports.LogLevel[level].toLowerCase().padEnd(LEVEL_MAX_LENGTH, " ");
            return COLOR_MAP[level](label);
        }(level)} -`;
        log.loggerLevel <= exports.LogLevel.Debug && (content = `${pc__default.default.gray(`[${function(time) {
            let h = time.getHours().toString().padStart(2, "0"), m = time.getMinutes().toString().padStart(2, "0"), s = time.getSeconds().toString().padStart(2, "0"), ms = time.getMilliseconds().toString().padStart(3, "0");
            return `${h}:${m}:${s}.${ms}`;
        }(time)}]`)} ${content}`), message ? content += ` ${message}` : commonUtils.isRecord(error) && "string" == typeof error.message && (content += ` ${error.message}`), null != data && (content += ` ${formatData(data)}`), error && (content += `\n${pc__default.default.gray(function(err) {
            if ("string" == typeof err) return err;
            if (commonUtils.isRecord(err)) {
                if ("string" == typeof err.stack) return cleanStack__default.default(err.stack, {
                    pretty: !0
                });
                if ("string" == typeof err.message) return err.message;
            }
            return formatData(err);
        }(error))}`), node_process.stderr.write(content + "\n");
    }
}

const createLogger = createLoggerFactory(()=>new NodeLogWriter());

class ConsoleLogWriter {
    write(log) {
        let args = [
            log.message,
            log.data,
            log.error
        ].filter(Boolean);
        switch(log.level){
            case exports.LogLevel.Trace:
                console.trace(...args);
                break;
            case exports.LogLevel.Debug:
                console.debug(...args);
                break;
            case exports.LogLevel.Info:
                console.info(...args);
                break;
            case exports.LogLevel.Warn:
                console.warn(...args);
                break;
            case exports.LogLevel.Error:
            case exports.LogLevel.Fatal:
                console.error(...args);
        }
    }
}

class SilentLogWriter {
    write() {}
}

var index = createLogger();

exports.ConsoleLogWriter = ConsoleLogWriter;
exports.Logger = Logger;
exports.SilentLogWriter = SilentLogWriter;
exports.createLogger = createLogger;
exports.default = index;
exports.logLevelFromString = logLevelFromString;
