// @kosko/env/dist/index.node.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var deepMerge = require('deepmerge');
var isPlainObject = require('is-plain-object');
var commonUtils = require('@kosko/common-utils');
var require$1 = require('@kosko/require');
var node_path = require('node:path');
var node_process = require('node:process');
var logger = require('@kosko/log');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var deepMerge__default = /*#__PURE__*/_interopDefault(deepMerge);
var logger__default = /*#__PURE__*/_interopDefault(logger);

function reduce(reducers, componentName) {
    return reducers.reduce((target, reducer)=>reducer.reduce(target, componentName), {});
}
async function reduceAsync(reducers, componentName) {
    let target = {};
    for (let reducer of reducers)target = await reducer.reduce(target, componentName);
    return target;
}

function merge(data) {
    return deepMerge__default.default.all(data, {
        isMergeableObject: isPlainObject.isPlainObject
    });
}
async function mergeAsync(data) {
    let values = await Promise.all(data);
    return merge(values);
}

function createEnvironment({ cwd ="/" , extensions =[] , setReducers , resetReducers , reduce  }) {
    return {
        cwd,
        paths: {
            global: "environments/#{environment}",
            component: "environments/#{environment}/#{component}"
        },
        extensions,
        setReducers,
        resetReducers,
        global: ()=>reduce(),
        component: (name)=>reduce(name)
    };
}
function createLoaderReducers(loader, mergeValues) {
    return [
        {
            name: "global",
            reduce: (values)=>mergeValues([
                    values,
                    ...commonUtils.toArray(loader.global())
                ])
        },
        {
            name: "component",
            reduce: (values, componentName)=>componentName ? mergeValues([
                    values,
                    ...commonUtils.toArray(loader.component(componentName))
                ]) : values
        }
    ];
}
function createSyncLoaderReducers(loader) {
    return createLoaderReducers(loader, merge);
}
function createAsyncLoaderReducers(loader) {
    return createLoaderReducers(loader, mergeAsync);
}
function createReducerList(initialReducers = []) {
    let reducers = [];
    function resetReducers() {
        reducers = [
            ...initialReducers
        ];
    }
    return resetReducers(), {
        getReducers: ()=>reducers,
        setReducers (callback) {
            reducers = callback([
                ...reducers
            ]);
        },
        resetReducers
    };
}
function createSyncReducerExecutor({ getReducers  }) {
    return {
        reduce: (componentName)=>reduce(getReducers(), componentName)
    };
}
function createAsyncReducerExecutor({ getReducers  }) {
    return {
        reduce: (componentName)=>reduceAsync(getReducers(), componentName)
    };
}

function createAsyncEnvironment() {
    let reducers = createReducerList(), { reduce  } = createAsyncReducerExecutor(reducers);
    return createEnvironment({
        ...reducers,
        reduce
    });
}

function createSyncEnvironment() {
    let reducers = createReducerList(), { reduce  } = createSyncReducerExecutor(reducers);
    return createEnvironment({
        ...reducers,
        reduce
    });
}

let rTemplate = /#\{(\w+)\}/g;
function formatPath(path, data) {
    return path.replace(rTemplate, (s, key)=>data[key] || s);
}

function getNodeExtensions() {
    return require$1.getRequireExtensions().map((ext)=>ext.substring(1));
}
function createNodeEnvironment({ cwd =node_process.cwd() , createReducerExecutor , requireModule , mergeValues  }) {
    function requireAllEnvs(template, component) {
        if (!environment.env) return [];
        let envs = commonUtils.toArray(environment.env);
        return envs.map((env)=>{
            let path = formatPath(template, {
                environment: env,
                ...component && {
                    component
                }
            });
            return requireModule(environment, node_path.join(environment.cwd, path));
        });
    }
    let reducers = createReducerList(createLoaderReducers({
        global: ()=>requireAllEnvs(environment.paths.global),
        component: (name)=>requireAllEnvs(environment.paths.component, name)
    }, mergeValues)), { reduce  } = createReducerExecutor(reducers), environment = createEnvironment({
        cwd,
        extensions: getNodeExtensions(),
        setReducers: reducers.setReducers,
        resetReducers: reducers.resetReducers,
        reduce
    });
    return environment;
}

function createNodeCJSEnvironment(options = {}) {
    return createNodeEnvironment({
        ...options,
        createReducerExecutor: createSyncReducerExecutor,
        mergeValues: merge,
        requireModule (env, id) {
            try {
                let mod = require(id);
                return mod && mod.__esModule ? mod.default : mod;
            } catch (err) {
                if ("MODULE_NOT_FOUND" === commonUtils.getErrorCode(err)) return logger__default.default.log(logger.LogLevel.Debug, `Cannot find module: ${id}`, {
                    error: err
                }), {};
                throw err;
            }
        }
    });
}

function createNodeESMEnvironment(options = {}) {
    return createNodeEnvironment({
        ...options,
        createReducerExecutor: createAsyncReducerExecutor,
        mergeValues: mergeAsync,
        async requireModule (env, id) {
            let path = await require$1.resolve(id, {
                extensions: env.extensions.map((ext)=>`.${ext}`)
            });
            if (!path) return logger__default.default.log(logger.LogLevel.Debug, "Module not found", {
                data: {
                    path: id,
                    extensions: env.extensions
                }
            }), {};
            logger__default.default.log(logger.LogLevel.Debug, `Importing ${path}`);
            let mod = await require$1.importPath(path);
            return mod.default;
        }
    });
}

const env = createNodeCJSEnvironment();

exports.createAsyncEnvironment = createAsyncEnvironment;
exports.createAsyncLoaderReducers = createAsyncLoaderReducers;
exports.createNodeCJSEnvironment = createNodeCJSEnvironment;
exports.createNodeESMEnvironment = createNodeESMEnvironment;
exports.createSyncEnvironment = createSyncEnvironment;
exports.createSyncLoaderReducers = createSyncLoaderReducers;
exports.default = env;
