// kubernetes-models/networking.k8s.io/v1beta1/IngressClass.mjs
import { createTypeMetaGuard, Model, setSchema } from "@kubernetes-models/base";
import { addSchema } from "../../_schemas/IoK8sApiNetworkingV1beta1IngressClass.mjs";
/**
 * IngressClass represents the class of the Ingress, referenced by the Ingress Spec. The `ingressclass.kubernetes.io/is-default-class` annotation can be used to indicate that an IngressClass should be considered default. When a single IngressClass resource has this annotation set to true, new Ingress resources without a class specified will be assigned this default class.
 */
export class IngressClass extends Model {
    constructor(data) {
        super({
            apiVersion: IngressClass.apiVersion,
            kind: IngressClass.kind,
            ...data
        });
    }
}
IngressClass.apiVersion = "networking.k8s.io/v1beta1";
IngressClass.kind = "IngressClass";
IngressClass.is = createTypeMetaGuard(IngressClass);
setSchema(IngressClass, "io.k8s.api.networking.v1beta1.IngressClass", addSchema);
export { IngressClass as IoK8sApiNetworkingV1beta1IngressClass };
