// kubernetes-models/storage.k8s.io/v1beta1/StorageClass.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IoK8sApiStorageV1beta1StorageClass = exports.StorageClass = void 0;
const base_1 = require("@kubernetes-models/base");
const IoK8sApiStorageV1beta1StorageClass_1 = require("../../_schemas/IoK8sApiStorageV1beta1StorageClass.js");
/**
 * StorageClass describes the parameters for a class of storage for which PersistentVolumes can be dynamically provisioned.
 *
 * StorageClasses are non-namespaced; the name of the storage class according to etcd is in ObjectMeta.Name.
 */
class StorageClass extends base_1.Model {
    constructor(data) {
        super({
            apiVersion: StorageClass.apiVersion,
            kind: StorageClass.kind,
            ...data
        });
    }
}
exports.StorageClass = StorageClass;
exports.IoK8sApiStorageV1beta1StorageClass = StorageClass;
StorageClass.apiVersion = "storage.k8s.io/v1beta1";
StorageClass.kind = "StorageClass";
StorageClass.is = (0, base_1.createTypeMetaGuard)(StorageClass);
(0, base_1.setSchema)(StorageClass, "io.k8s.api.storage.v1beta1.StorageClass", IoK8sApiStorageV1beta1StorageClass_1.addSchema);
