// kubernetes-models/networking.k8s.io/v1/IngressClass.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IoK8sApiNetworkingV1IngressClass = exports.IngressClass = void 0;
const base_1 = require("@kubernetes-models/base");
const IoK8sApiNetworkingV1IngressClass_1 = require("../../_schemas/IoK8sApiNetworkingV1IngressClass.js");
/**
 * IngressClass represents the class of the Ingress, referenced by the Ingress Spec. The `ingressclass.kubernetes.io/is-default-class` annotation can be used to indicate that an IngressClass should be considered default. When a single IngressClass resource has this annotation set to true, new Ingress resources without a class specified will be assigned this default class.
 */
class IngressClass extends base_1.Model {
    constructor(data) {
        super({
            apiVersion: IngressClass.apiVersion,
            kind: IngressClass.kind,
            ...data
        });
    }
}
exports.IngressClass = IngressClass;
exports.IoK8sApiNetworkingV1IngressClass = IngressClass;
IngressClass.apiVersion = "networking.k8s.io/v1";
IngressClass.kind = "IngressClass";
IngressClass.is = (0, base_1.createTypeMetaGuard)(IngressClass);
(0, base_1.setSchema)(IngressClass, "io.k8s.api.networking.v1.IngressClass", IoK8sApiNetworkingV1IngressClass_1.addSchema);
