// kubernetes-models/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.flowcontrolApiserverK8sIo = exports.discoveryK8sIo = exports.apiregistrationK8sIo = exports.apiextensionsK8sIo = exports.storageK8sIo = exports.schedulingK8sIo = exports.rbacAuthorizationK8sIo = exports.policy = exports.nodeK8sIo = exports.networkingK8sIo = exports.extensions = exports.eventsK8sIo = exports.v1 = exports.coordinationK8sIo = exports.certificatesK8sIo = exports.batch = exports.autoscaling = exports.authorizationK8sIo = exports.authenticationK8sIo = exports.apps = exports.admissionregistrationK8sIo = void 0;
const tslib_1 = require("tslib");
exports.admissionregistrationK8sIo = tslib_1.__importStar(require("./admissionregistration.k8s.io/index.js"));
exports.apps = tslib_1.__importStar(require("./apps/index.js"));
exports.authenticationK8sIo = tslib_1.__importStar(require("./authentication.k8s.io/index.js"));
exports.authorizationK8sIo = tslib_1.__importStar(require("./authorization.k8s.io/index.js"));
exports.autoscaling = tslib_1.__importStar(require("./autoscaling/index.js"));
exports.batch = tslib_1.__importStar(require("./batch/index.js"));
exports.certificatesK8sIo = tslib_1.__importStar(require("./certificates.k8s.io/index.js"));
exports.coordinationK8sIo = tslib_1.__importStar(require("./coordination.k8s.io/index.js"));
exports.v1 = tslib_1.__importStar(require("./v1/index.js"));
exports.eventsK8sIo = tslib_1.__importStar(require("./events.k8s.io/index.js"));
exports.extensions = tslib_1.__importStar(require("./extensions/index.js"));
exports.networkingK8sIo = tslib_1.__importStar(require("./networking.k8s.io/index.js"));
exports.nodeK8sIo = tslib_1.__importStar(require("./node.k8s.io/index.js"));
exports.policy = tslib_1.__importStar(require("./policy/index.js"));
exports.rbacAuthorizationK8sIo = tslib_1.__importStar(require("./rbac.authorization.k8s.io/index.js"));
exports.schedulingK8sIo = tslib_1.__importStar(require("./scheduling.k8s.io/index.js"));
exports.storageK8sIo = tslib_1.__importStar(require("./storage.k8s.io/index.js"));
exports.apiextensionsK8sIo = tslib_1.__importStar(require("./apiextensions.k8s.io/index.js"));
exports.apiregistrationK8sIo = tslib_1.__importStar(require("./apiregistration.k8s.io/index.js"));
exports.discoveryK8sIo = tslib_1.__importStar(require("./discovery.k8s.io/index.js"));
exports.flowcontrolApiserverK8sIo = tslib_1.__importStar(require("./flowcontrol.apiserver.k8s.io/index.js"));
